src/lib/tp.c

```diff
@@ -0,0 +1,152 @@
+/*
+ *   Software Updater - client side
+ *
+ *      Copyright © 2018 Intel Corporation.
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *   Authors:
+ *         Otavio Pontes <otavio.pontes@intel.com>
+ *
+ */
+
+#define _GNU_SOURCE
+
+#include <errno.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "macros.h"
+#include "tp.h"
+
+/*
+ * To keep this thread pool implementation simple, pipes were used instead of
+ * mutex locks using the fact that read() and write() operations are thread safe.
+ * So in order to schedule a task to be executed we just need to write() a task
+ * in to the pipe and the first thread to read() that data will execute the
+ * processing.
+ */
+
+struct tp {
+	int num_threads;
+	int pipe_fds[2];
+	pthread_t threads[];
+};
+
+struct task {
+	tp_task_run_t run;
+	void *data;
+};
+
+static void *thread_run(void *data)
+{
+	int *fd;
+	ssize_t r;
+	struct task task;
+
+	fd = data;
```

Comment:
You could check that `*fd` is actually a valid integer > 0 before trying to read anything here, and exit early if it's not.